--no-add installs only specified specs and only if already present in…

scottwittenburg opened the pull request:
… env

Fixes #23333

Allow installing one or more specs from command line as long as they are
already present in the environment.  ~~Change spec parsing to check active
environment for dag hash (in addition to checking installation database).~~

~~Instead of changing spec parsing, this now relies on #23194 for that functionality.~~

This is now decoupled from parsing hashes of env specs on the command line, and thus no longer requires #23194.

haampie commented:
I didn't thoroughly check, but are you aware of `spec = spack.cmd.matching_spec_from_env(spec)` which got added to a few commands (`build-env`, `stage`, `location`)? Looks similar

scottwittenburg commented:
This is my attempt to implement #22586.

@tgamblin I think this is more or less what we talked about last Friday.  If you know of the right reviewer, feel free to add them.

scottwittenburg commented:
> I didn't thoroughly check, but are you aware of `spec = spack.cmd.matching_spec_from_env(spec)` which got added to a few commands (`build-env`, `stage`, `location`)? Looks similar

Well I did try the method that calls in the `Environment` class (which is called `matching_spec`), but when used on the hash of a dependency in the environment, it fails and prints a bunch of messages that the spec in question is a dependency of a bunch of other specs.  Otherwise, I could have used that method to find the target for installation.  Maybe I just mis-used it though?

EDIT: Having looked at this carefully this week, I do think we should use `env.matching_spec(spec)` from the `Environment` class, due to how it intends to handle when a spec might match `n` roots and `m` deps (for `n >= 0` and `m>=0`).  However I ran into a couple of issues (one confirmed bug, one questioned) and wrote them up in #23333.  Depending on the resolution of the questioned bug (`matching_spec` can return an abstract spec rather than the concrete one from the environment), I plan to move forward using that method instead of searching through the specs myself.

scottwittenburg commented:
fyi @zackgalbreath

hartzell commented:
@scottwittenburg -- just wanted to drop a note here to follow up on using this work to leverage being able to use the CI pipeline to build buildcaches for environments that are concretized together.

scheibelp committed:
fix check when retrieving matching spec from environment when there is a match with a root spec as well as with a dependency of a root spec

scheibelp committed:
return concrete spec (as is advertised in the docstring)

scottwittenburg committed:
install cmd: --no-add in an env installs w/out concretize and add

In an active concretize environment, support installing one or more
cli specs only if they are already present in the environment.  The
`--no-add` option is the default for root specs, but optional for
dependency specs.  I.e. if you `spack install <depspec>` in an
environment, the dependency-only spec `depspec` will be added as a
root of the environment before being installed.  In addition,
`spack install --no-add <spec>` fails if it does not find an
unambiguous match for `spec`.

scottwittenburg commented:
@hartzell I think you should be able to use this branch now to `spack install --no-add <spec>` in an environment concretized `together`.  The only requirement is that each spec you provide on the command line needs to match unambiguously a spec already added to the environment.  

As I mentioned in the description, I decoupled this PR from #23194, which allows parsing hashes on the command line for uninstalled specs spack has "seen" before (e.g. specs in active env or in binary mirrors).  Consequently, it's not as easy to unambiguously specify a spec on the command line.  Until we get that work in, my plan in CI is to use concrete spec yaml files written to disk to specify target specs unambiguously, as I have done all along.  That's not strictly required though, abstract cli specs are just fine as long as they match exactly one spec in the active environment.

scottwittenburg commented:
@tgamblin I think this is ready, can you assign a reviewer based on availability, etc?  I think codecov is just wrong about the patch coverage, it's actually more like 100%, and on top of that codecov is complaining about uncovered lines in a module not even touched by this PR (`lib/spack/spack/util/compression.py`).

haampie commented:
Some commands have been made "environment aware" recently, including `spack -e env build-env [spec]`, `spack -e env stage [spec]`, and more. They filter concrete matching specs in the environment, and then apply `stage` or `build-env` to them.

This makes me wonder: shouldn't the behavior of `--no-add` be the default for consistency with the above?

I would be happy if this was the default behavior in environments:

```
spack -e env add <spec> # adds a non-concrete root spec <spec> without installing it
spack -e env install # installs the full environment
spack -e env install <spec> # installs all specs in the concrete environment matching <spec> or none if there is no match
spack -e env uninstall <spec> # uninstalls all specs in the concrete environment matching <spec> -- also this shouldn't remove root specs from the environment file!!
spack -e env uninstall # uninstalls the full environment without removing root specs from the environment file
spack -e env remove <spec> # removes matching root specs from the environment
```

scottwittenburg commented:
@haampie I kind of agree with your assessment there, at least from my narrow perspective which is mostly focused on gitlab CI pipelines.  However, I've heard @tgamblin and @becker33 express the argument that spack environments should behave similar to python virtual environments (so `spack install foo` should add `foo` to the environment).

haampie commented:
Yeah, `spack -e env install xyz` could still add a new root spec if there is no matching concrete spec in the environment.

In fact the current behavior of `spack -e env build-env xyz` is this:
- if xyz matches a single concrete spec, give the build-env of that
- if xyz does not match any concrete spec, concretize xyz in the context of the environment, and give the build-env of that

So the same could apply to `spack install`, right?
- if xyz matches one or more concrete specs, install all of these
- if there is no match, run spack add xyz, concretize, and "rerun" spack install xyz

scottwittenburg commented:
> So the same could apply to `spack install`, right?
> 
> * if xyz matches one or more concrete specs, install all of these
> * if there is no match, run spack add xyz, concretize, and "rerun" spack install xyz

I'm inclined to agree with you here.  Maybe I'm missing something, but this makes the most sense to me.

adrienbernede commented:
Reading this discussion:
- Environment are designed to mimic python environment behavior: add a top-level requirement, dependencies are implied.
- Environments will not allow independent builds of the dependencies without adding it to the top-level spec (hence this PR).
- However, even outside CI, there could be a need to install a given package in the context of an environment (without altering the environment definition (think facility software stacks: an environment describes all the vetted specs)).

This made me think that there is a need for "contexts" (sorts of 'read-only' environments). I am satisfied with the way this PR allows me to use an environment as a "context" description.

scottwittenburg commented:
> So the same could apply to `spack install`, right?
> 
> * if xyz matches one or more concrete specs, install all of these
> * if there is no match, run spack add xyz, concretize, and "rerun" spack install xyz

scottwittenburg commented:
Sorry, accidental PR close there.

scottwittenburg commented:
@tgamblin @becker33 Do either of you see any reason not to make this behavior the default (and consequently do away with the `--no-add` option) as suggested in the [comment](https://github.com/spack/spack/pull/22657#issuecomment-832851432) by @haampie?

tgamblin reviewed (approved):
I think we want to keep the default venv behavior of adding for now, but I agree it's sort of a mixed metaphor.  I think this is good as-is, but there probably need to be a larger UI/UX investigation around environments and all the different use cases.

The issue is that they satisfy a *lot* of use cases.  I don't think we should change the default install right now without considering how that affects them, and I think this PR at least gets us to where we can do what @haampie, @adrienbernede, and others want.
